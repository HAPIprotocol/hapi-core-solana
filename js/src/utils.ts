import BN from "bn.js";
import { decode } from "bs58";
import {
  Connection,
  Keypair,
  PublicKey,
  Transaction,
  TransactionInstruction,
} from "@solana/web3.js";

export class u8 extends BN {
  static size = 1;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u8.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u8.size);
  }

  /**
   * Construct a Numberu8 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u8.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(4)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u32 extends BN {
  static size = 4;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u32.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u32.size);
  }

  /**
   * Construct a Numberu32 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u32.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u64 extends BN {
  static size = 8;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u64.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u64.size);
  }

  /**
   * Construct a Numberu64 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u64.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export function bnToBuffer(bn: BN, size: number): Buffer {
  const a = bn.toArray().reverse();
  const b = Buffer.from(a);
  if (b.length === size) {
    return b;
  }
  if (b.length > size) {
    throw new Error(`Buffer too large (size: ${size})`);
  }

  const zeroPad = Buffer.alloc(size);
  b.copy(zeroPad);
  return zeroPad;
}

export const signAndSendTransactionInstructions = async (
  // sign and send transaction
  connection: Connection,
  signers: Array<Keypair>,
  feePayer: Keypair,
  txInstructions: Array<TransactionInstruction>
): Promise<string> => {
  const tx = new Transaction();
  tx.feePayer = feePayer.publicKey;
  signers.push(feePayer);
  tx.add(...txInstructions);
  return await connection.sendTransaction(tx, signers, {
    preflightCommitment: "single",
  });
};

export function mapToBuffer<K extends number, V extends number | boolean>(
  map: Map<K, V>
): Buffer {
  const buffer = Buffer.alloc(4 + map.size * 2);

  // Write size header
  buffer.writeUInt32LE(map.size);

  // Write map KVs
  let i = 4;
  const keys = Array.from(map.keys()).sort((a, b) => a - b);
  for (const key of keys) {
    buffer.writeUInt8(key, i);
    buffer.writeUInt8(map.get(key) ? 1 : 0, i + 1);
    i += 2;
  }

  return buffer;
}

export function setToBuffer<T extends number>(set: Set<T>, size = 4): Buffer {
  let value = new BN(0);

  for (const i of set.values()) {
    value = value.or(new BN(i));
  }

  return value.toBuffer("le", size);
}

export function base58ToPublicKey(address: string): PublicKey {
  const buffer = decode(address);
  return new PublicKey(buffer);
}

export function hexToPublicKey(hex: string): PublicKey {
  const byteLength = Buffer.from(hex, "hex").length;
  if (byteLength > 32) {
    throw RangeError(
      `hexToPublicKey: input length can't be longer than 32 bytes`
    );
  }
  const buffer = Buffer.alloc(32);
  buffer.write(hex, 32 - byteLength, "hex");
  return PublicKey.decodeUnchecked(buffer);
}
