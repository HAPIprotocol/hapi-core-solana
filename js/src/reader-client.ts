import {
  Commitment,
  Connection,
  ConnectionConfig,
  PublicKey,
} from "@solana/web3.js";

import { base58ToPublicKey, u64 } from "./utils";
import { Address, Case, Community, Network, Reporter } from "./state";
import { HAPI_PROGRAM_ID } from "./constants";

export interface HapiClientConfig {
  endpoint: string;
  commintment?: Commitment | ConnectionConfig;
  communityName?: string;
  programId?: string;
}

export interface HapiViewResponse<T> {
  account: PublicKey;
  data: T;
}

export interface HapiActionResponse<T> {
  account: PublicKey;
  txHash: string;
  data: T;
}

/** HAPI client to read entity data from Solana */
export class ReaderClient {
  protected connection: Connection;
  protected communityName?: string;
  protected programId: PublicKey;

  constructor(config: HapiClientConfig) {
    this.connection = new Connection(config.endpoint, config.commintment);
    this.communityName = config.communityName;
    this.programId = new PublicKey(this.programId || HAPI_PROGRAM_ID);
  }

  /**
   * Sets community name for context
   * @param communityName Community name
   * @returns Self
   */
  switchCommunity(communityName: string): ReaderClient {
    this.communityName = communityName;
    return this;
  }

  /**
   * Fetch community info from blockchain
   * @param communityName (Optional) The name of the community to fetch (defaults to context)
   * @returns Community info
   **/
  async getCommunity(
    communityName?: string
  ): Promise<HapiViewResponse<Community>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Community.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName
    );

    return state;
  }

  /**
   * Fetch network info from blockchain
   * @param networkName The name of the network to fetch
   * @param communityName (Optional) The name of the community to fetch network from (defaults to context)
   * @returns Network info
   **/
  async getNetwork(
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Network>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const state = await Network.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName
    );

    return state;
  }

  /**
   * Fetch reporter info from blockchain
   * @param reporterPubkey Public key of the reporter to fetch
   * @param communityName (Optional) The name of the community to fetch reporter from (defaults to context)
   * @returns Reporter info
   **/
  async getReporter(
    reporterPubkey: string,
    communityName?: string
  ): Promise<HapiViewResponse<Reporter>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Reporter.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      new PublicKey(reporterPubkey)
    );

    return state;
  }

  /**
   * Fetch case info from blockchain
   * @param caseId ID of the case to fetch
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Case info
   **/
  async getCase(
    caseId: u64,
    communityName?: string
  ): Promise<HapiViewResponse<Case>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Case.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      caseId
    );

    return state;
  }

  /**
   * Fetch address info from blockchain
   * @param address The address to fetch info for (string for Solana addresses, Buffer for others)
   * @param networkName The name of the network to which address belongs to
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Address info
   **/
  async getAddress(
    address: string | Buffer,
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Address>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const convertedAddress =
      address instanceof Buffer
        ? new PublicKey(address)
        : base58ToPublicKey(address);

    const state = await Address.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName,
      convertedAddress
    );

    return state;
  }
}
