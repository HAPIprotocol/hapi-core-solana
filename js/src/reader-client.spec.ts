import bs58 from "bs58";
import nock from "nock";

import { ReaderClient } from "./reader-client";
import { u64 } from "./utils";
import { mockRpcOk } from "../test/util/mocks";

describe("ReaderClient", () => {
  let client: ReaderClient;

  const endpoint = "http://localhost:8899";

  beforeAll(() => {
    nock.disableNetConnect();
  });

  beforeEach(() => {
    nock.cleanAll();
    client = new ReaderClient({ endpoint });
  });

  it("should initialize", () => {
    expect(client).toBeDefined();
  });

  describe("getCommunity", () => {
    it("should throw - not found", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh"],
        { context: { slot: 173 }, value: null }
      );

      await expect(() =>
        client.getCommunity("community404")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja"],
        {
          context: { slot: 280 },
          value: {
            data: [
              "Ae83pQQEsYipbhD1URInU/iuiQbErgmDyScnrPJbOPlRAgAAAAAAAAAIAAAAaGFwaS5vbmUAAAAAAAAAAAAAAAAAAAAAAAAAAA===",
              "base64",
            ],
            executable: false,
            lamports: 1398960,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const response = await client.getCommunity("hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("getNetwork", () => {
    it("should throw - not found", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["3TGzDJHkCHX5R62wtVh1Jr5TNe775jvSCt8tThbizBKQ"],
        { context: { slot: 4682 }, value: null }
      );

      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja"],
        {
          context: { slot: 280 },
          value: {
            data: [
              "Ae83pQQEsYipbhD1URInU/iuiQbErgmDyScnrPJbOPlRAgAAAAAAAAAIAAAAaGFwaS5vbmUAAAAAAAAAAAAAAAAAAAAAAAAAAA===",
              "base64",
            ],
            executable: false,
            lamports: 1398960,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["7GiNHuwBKg6URcBJXXXRPRtBVgn5q42AiowBEc8uzqBH"],
        { context: { slot: 4682 }, value: null }
      );

      await expect(() =>
        client.getNetwork("unnetwork", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();

      await expect(() =>
        client.getNetwork("testcoin", "community404")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["2viJmmn2pEfd6cogyqdDGWS9YkrVdnx87L994Qo3GwLx"],
        {
          context: { slot: 4682 },
          value: {
            data: ["AggAAAB0ZXN0Y29pbgAAAAAAAAAAAAAAAAAAAAAAAAAA", "base64"],
            executable: false,
            lamports: 1120560,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const response = await client.getNetwork("testcoin", "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("getReporter", () => {
    it("should throw - not found", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["DUALaTZpDYtct73t8D9AirPYnPc4TzSCmqwcNjxmSf9j"],
        { context: { slot: 5254 }, value: null }
      );

      await expect(() =>
        client.getReporter(
          "49B9UtxhbVBKaToxThwQqshbmf2ZfnuJZX3XGjTsNLvb",
          "hapi.one"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      // Alice
      {
        mockRpcOk(
          endpoint,
          "getAccountInfo",
          ["GfwYi1NaoMFJUHzEXtTkXAQewxxqs7PbseAYnsfiNnS7"],
          {
            context: { slot: 5254 },
            value: {
              data: [
                "AwMFAAAAQWxpY2UAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
                "base64",
              ],
              executable: false,
              lamports: 1127520,
              owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
              rentEpoch: 0,
            },
          }
        );

        const response = await client.getReporter(
          "DzMkTkH6ms7hEzyHisFnLLc2WDJfBb9TNNaPDQ7ADHhy",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }

      // Bob
      {
        mockRpcOk(
          endpoint,
          "getAccountInfo",
          ["7GFtavi6PpFjqmvhKMtW8D6W4JgFv69VhUurrnBQRbUL"],
          {
            context: { slot: 5254 },
            value: {
              data: [
                "AwIDAAAAQm9iAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
                "base64",
              ],
              executable: false,
              lamports: 1127520,
              owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
              rentEpoch: 0,
            },
          }
        );

        const response = await client.getReporter(
          "J5sUnZKuB1a9izNWDb4JEQzdB3J6mhe3sP6Ai6YCiKAZ",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }

      // Carol
      {
        mockRpcOk(
          endpoint,
          "getAccountInfo",
          ["DSa28TFYUaomAUNTnRRcsxxmZmvrf1Pfim7obB21m5Jx"],
          {
            context: { slot: 5254 },
            value: {
              data: [
                "AwEFAAAAQ2Fyb2wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
                "base64",
              ],
              executable: false,
              lamports: 1127520,
              owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
              rentEpoch: 0,
            },
          }
        );

        const response = await client.getReporter(
          "E8E298Dfe79V97ngYkcxXo1BftwduxjkGUWWRePM1Ac2",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }
    });
  });

  describe("getCase", () => {
    it("should throw - not found", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["EGP3s7nD3dFaT9jGtwT7UoZndi58W3VexaJ41N1y5yMm"],
        {
          context: { slot: 8869 },
          value: {
            data: ["BQUBAAAAAAAAAA8=", "base64"],
            executable: false,
            lamports: 967440,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["EGP3s7nD3dFaT9jGtwT7UoZndi58W3VexaJ41N1y5yMm"],
        {
          context: { slot: 8869 },
          value: {
            data: ["BQUBAAAAAAAAAA8=", "base64"],
            executable: false,
            lamports: 967440,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["5yGEyZT2WwuLD6BJuvP7TXRcR2ojZTpFVv2TkQm6eX2P"],
        { context: { slot: 5254 }, value: null }
      );
      await expect(() =>
        client.getCase(new u64(404), "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["63G8TLWGQpd26UZj7L9Qr9e3R1MPbybLcW3A7LXtG1Sk"],
        {
          context: { slot: 10001 },
          value: {
            data: [
              "BMD9z4HkaJp54Mtk2ICY9TQpEGUqNA3cBwPb2xA4bcZ4AAAAAAAFAAAAY2FzZTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
              "base64",
            ],
            executable: false,
            lamports: 1385040,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const response = await client.getCase(new u64(0), "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("getAddress", () => {
    it("should throw - not found (string)", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["8HpP3NYL4x8qifD3g2e1z7M1m4wPNeXtMoT8Z7yt1Yzs"],
        { context: { slot: 6486 }, value: null }
      );

      await expect(() =>
        client.getAddress("4o4", "testcoin", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - not found (buffer)", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["7viPn64FE3Uz9bFqKYDnBsL4ndwb8rXhFPjZ9NvgCVSx"],
        { context: { slot: 6911 }, value: null }
      );

      await expect(() =>
        client.getAddress(
          Buffer.from("deadcode", "hex"),
          "testcoin",
          "hapi.one"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - malformed", async () => {
      await expect(() =>
        client.getAddress("BAD BASE58", "testcoin", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid community name", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["BMCaLwP7mmErK69cay99uSnyV1YRtFGGTRU8Vj4F7KNd"],
        { context: { slot: 6984 }, value: null }
      );

      await expect(() =>
        client.getAddress(
          "2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew",
          "testcoin",
          "community404"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success (string)", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["C5Srms9M3eHy5XbzEgRgEEHFZ3rKxdAEaH2DZSSXLGDD"],
        {
          context: { slot: 7526 },
          value: {
            data: ["BQUBAAAAAAAAAA8=", "base64"],
            executable: false,
            lamports: 967440,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const response = await client.getAddress(
        "2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew",
        "testcoin",
        "hapi.one"
      );
      expect(response).toMatchSnapshot();
    });

    it("should respond - success (buffer)", async () => {
      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["C5Srms9M3eHy5XbzEgRgEEHFZ3rKxdAEaH2DZSSXLGDD"],
        {
          context: { slot: 7526 },
          value: {
            data: ["BQUBAAAAAAAAAA8=", "base64"],
            executable: false,
            lamports: 967440,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const buffer = bs58.decode(
        "2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew"
      );
      const response = await client.getAddress(buffer, "testcoin", "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });
});
