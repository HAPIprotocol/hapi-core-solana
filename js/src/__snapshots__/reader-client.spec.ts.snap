// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReaderClient getAddress should respond - success (buffer) 1`] = `
Object {
  "account": PublicKey {
    "_bn": "c51892d0703b31cdf66396ec41f908351cfda218f07c6f5350d767bf040654f4",
  },
  "data": Address {
    "accountType": 5,
    "caseId": "01",
    "category": 16384,
    "risk": 5,
  },
}
`;

exports[`ReaderClient getAddress should respond - success (string) 1`] = `
Object {
  "account": PublicKey {
    "_bn": "c51892d0703b31cdf66396ec41f908351cfda218f07c6f5350d767bf040654f4",
  },
  "data": Address {
    "accountType": 5,
    "caseId": "01",
    "category": 16384,
    "risk": 5,
  },
}
`;

exports[`ReaderClient getAddress should throw - invalid community name 1`] = `"Address not found: \\"2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew\\" in network \\"testcoin\\" (7GiNHuwBKg6URcBJXXXRPRtBVgn5q42AiowBEc8uzqBH) in community \\"community404\\" (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient getAddress should throw - malformed 1`] = `"Non-base58 character"`;

exports[`ReaderClient getAddress should throw - not found (buffer) 1`] = `"Address not found: \\"111111111111111111111111111111Hwr\\" in network \\"testcoin\\" (2viJmmn2pEfd6cogyqdDGWS9YkrVdnx87L994Qo3GwLx) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient getAddress should throw - not found (string) 1`] = `"Address not found: \\"1111111111111111111111111111114o4\\" in network \\"testcoin\\" (2viJmmn2pEfd6cogyqdDGWS9YkrVdnx87L994Qo3GwLx) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient getCase should respond - success 1`] = `
Object {
  "account": PublicKey {
    "_bn": "bc563659f0d64faa3745edf63ead612d47d3ae9f326724afc08388afbe0a90ef",
  },
  "data": Case {
    "accountType": 4,
    "categories": Array [],
    "name": "case0",
    "reporterKey": PublicKey {
      "_bn": "c0fdcf81e4689a79e0cb64d88098f5342910652a340ddc0703dbdb10386dc678",
    },
    "status": 0,
  },
}
`;

exports[`ReaderClient getCase should throw - not found 1`] = `"Invalid case account provided"`;

exports[`ReaderClient getCommunity should respond - success 1`] = `
Object {
  "account": PublicKey {
    "_bn": "bc563659f0d64faa3745edf63ead612d47d3ae9f326724afc08388afbe0a90ef",
  },
  "data": Community {
    "accountType": 1,
    "authority": PublicKey {
      "_bn": "ef37a50404b188a96e10f551122753f8ae8906c4ae0983c92727acf25b38f951",
    },
    "name": "hapi.one",
    "nextCaseId": "02",
  },
}
`;

exports[`ReaderClient getCommunity should throw - not found 1`] = `"Community not found: community404 (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient getNetwork should respond - success 1`] = `
Object {
  "account": PublicKey {
    "_bn": "1c9e2cd47719bf0f4982def71c6a39dff80f7eb1f30cf50d7c28e0a8b12ce295",
  },
  "data": Network {
    "accountType": 2,
    "name": "testcoin",
  },
}
`;

exports[`ReaderClient getNetwork should throw - not found 1`] = `"Network not found: \\"unnetwork\\" (3TGzDJHkCHX5R62wtVh1Jr5TNe775jvSCt8tThbizBKQ) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient getNetwork should throw - not found 2`] = `"Network not found: \\"testcoin\\" (7GiNHuwBKg6URcBJXXXRPRtBVgn5q42AiowBEc8uzqBH) in community \\"community404\\" (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient getReporter should respond - success 1`] = `
Object {
  "account": PublicKey {
    "_bn": "e8d90cfe9173301e5ca4c73615c4ff649b8599d5cd8ca4d370c5ac5ecb6ea59c",
  },
  "data": Reporter {
    "accountType": 3,
    "name": "Alice",
    "reporterType": 3,
  },
}
`;

exports[`ReaderClient getReporter should respond - success 2`] = `
Object {
  "account": PublicKey {
    "_bn": "5d0eab4fdb1135df313e7485d099bf91edec7576ca3f0943853a3c0d77b6de29",
  },
  "data": Reporter {
    "accountType": 3,
    "name": "Bob",
    "reporterType": 2,
  },
}
`;

exports[`ReaderClient getReporter should respond - success 3`] = `
Object {
  "account": PublicKey {
    "_bn": "b8d9193b5a70f5cf8932d373ec3e51abb1fe41d2f3a20ba3e579c18eaf1e12a1",
  },
  "data": Reporter {
    "accountType": 3,
    "name": "Carol",
    "reporterType": 1,
  },
}
`;

exports[`ReaderClient getReporter should throw - not found 1`] = `"Reporter not found: \\"49B9UtxhbVBKaToxThwQqshbmf2ZfnuJZX3XGjTsNLvb\\" in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;
